### PR TITLE
[target-allocator] Populate store assets (authorization information) for Prometheus CR watcher

### LINKED ISSUE #1669: Target Allocator - ServiceMonitor scheme
Hi

I have the TA up and with the `PrometheusCR` enabled, this is to look at options of migrating to using the OTEL collector for more purposes.

At present I am getting data as expected except for service monitors (to be fair in my test lab I only have a couple of these setup) that contain a "non default" scheme and have authentication data in them. eg

```
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  creationTimestamp: "2023-04-06T18:07:21Z"
  generation: 1
  labels:
    app: first-cluster
  name: first-cluster
  namespace: opensearch-first-cluster
  resourceVersion: "991661"
  uid: cfad90db-7361-4b05-96de-b89b36537dbc
spec:
  endpoints:
  - basicAuth:
      password:
        key: password
        name: first-cluster-opensearch-monitoring
      username:
        key: username
        name: first-cluster-opensearch-monitoring
    interval: 30s
    path: /_prometheus/metrics
    port: http
    scheme: https
    tlsConfig:
      insecureSkipVerify: true
  jobLabel: opensearch_first-cluster
  namespaceSelector:
    matchNames:
    - opensearch-first-cluster
  selector:
    matchLabels:
      opster.io/opensearch-cluster: first-cluster
```

```
apiVersion: v1
kind: Service
metadata:
  annotations:
    banzaicloud.com/last-applied: UEsDBBQACAAIAAAAAAAAAAAAAAAAAAAAAAAIAAAAb3JpZ2luYWyUks2O2zAMhN+FZ3nrrGO362O39wRN0UuRAy3TjRBFEig6PQR690JO4Py0QLA3ezTzYUjwBBjMT+JovIMWjgtQsDeuhxY2xEejCRQcSLBHQWhPYLEjG/OXD1GIX4z/5AO5SMh6V2g7ZhVaGAxHmf+TAocH+kc/yzGgzm83oEeb/+OIv9NATE5ThPbXQ/Vr9uXabJqns17vVzn/jSzJZBceSYH2Tthbmwuflcvsq0BuM8He74r+p/9ocqBrqhL1siwWr8NbsWw+98WXCrFoqnro+mW5oK6GtE0KYiCd1xc8y3mMC3gnEkBNOrRvr2WpILAXr72FFn68r0GBIP8mWc+WpOa4MLo4hWdG9ZxR3TEOJGx0vBKa54TmjsAab9L183Rdpq2CSJa0eP74XeWNCso4naT12H9Fi05n6yml9DcAAP//UEsHCDhHq9BCAQAA4AIAAFBLAQIUABQACAAIAAAAAAA4R6vQQgEAAOACAAAIAAAAAAAAAAAAAAAAAAAAAABvcmlnaW5hbFBLBQYAAAAAAQABADYAAAB4AQAAAAA=
  creationTimestamp: "2023-04-06T18:07:23Z"
  labels:
    opster.io/opensearch-cluster: first-cluster
  name: first-cluster
  namespace: opensearch-first-cluster
  ownerReferences:
  - apiVersion: opensearch.opster.io/v1
    blockOwnerDeletion: true
    controller: true
    kind: OpenSearchCluster
    name: first-cluster
    uid: b630ac40-12f9-467d-83aa-635fbd401eb5
  resourceVersion: "991678"
  uid: add9fb87-651a-4898-8a10-ccc29ed0b043
spec:
  clusterIP: 192.168.5.15
  clusterIPs:
  - 192.168.5.15
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - name: http
    port: 9200
    protocol: TCP
    targetPort: 9200
  - name: transport
    port: 9300
    protocol: TCP
    targetPort: 9300
  - name: metrics
    port: 9600
    protocol: TCP
    targetPort: 9600
  - name: rca
    port: 9650
    protocol: TCP
    targetPort: 9650
  selector:
    opster.io/opensearch-cluster: first-cluster
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
```

This config works with prometheus but using the otel collector (via the TA) I get `2023-04-20T20:04:56.248Z	warn	internal/transaction.go:121	Failed to scrape Prometheus endpoint	{"kind": "receiver", "name": "prometheus", "data_type": "metrics", "scrape_timestamp": 1682021096246, "target_labels": "{__name__=\"up\", container=\"opensearch\", endpoint=\"http\", instance=\"100.64.141.115:9200\", job=\"first-cluster-masters\", namespace=\"opensearch-first-cluster\", pod=\"first-cluster-masters-1\", service=\"first-cluster-masters\"}"}`

Just looking to find out if this is a known "issue" or is, and it probably is, it that I am doing something wrong? Any help is appreciated

This is all being deployed using the opentelemetry-operator helm chart version 0.26.3 

Target Allocator
```
apiVersion: v1
data:
  targetallocator.yaml: |
    allocation_strategy: least-weighted
    config:
      scrape_configs:
      - job_name: otel-collector
        scrape_interval: 10s
        static_configs:
        - targets:
          - 0.0.0.0:8888
    label_selector:
      app.kubernetes.io/component: opentelemetry-collector
      app.kubernetes.io/instance: otel.allocator
      app.kubernetes.io/managed-by: opentelemetry-operator
kind: ConfigMap
metadata:
  creationTimestamp: "2023-04-20T18:37:48Z"
  labels:
    app.kubernetes.io/component: opentelemetry-targetallocator
    app.kubernetes.io/instance: otel.allocator
    app.kubernetes.io/managed-by: opentelemetry-operator
    app.kubernetes.io/name: allocator-targetallocator
    app.kubernetes.io/part-of: opentelemetry
    app.kubernetes.io/version: latest
  name: allocator-targetallocator
  namespace: otel
  ownerReferences:
  - apiVersion: opentelemetry.io/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: OpenTelemetryCollector
    name: allocator
    uid: 012622e4-db73-4c9c-8f7c-39e47cdabb8f
  resourceVersion: "9749302"
  uid: a591df05-6f7e-4267-8953-338363472f87
```

Collector
```
apiVersion: v1
data:
  collector.yaml: |
    exporters:
      logging:
        loglevel: debug
      prometheusremotewrite:
        endpoint: http://prometheus.kube-prometheus-stack.svc.cluster.local:9090/api/v1/write
        external_labels:
          scraper: ${POD_NAME}
          source: otel-allocator
    extensions:
      memory_ballast:
        size_in_percentage: 20
    processors:
      batch:
        send_batch_max_size: 1000
        send_batch_size: 800
        timeout: 30s
      memory_limiter:
        check_interval: 1s
        limit_percentage: 70
        spike_limit_percentage: 30
    receivers:
      hostmetrics:
        collection_interval: 30s
        scrapers:
          cpu: null
          load:
            cpu_average: true
          memory: null
      prometheus:
        config:
          global:
            scrape_interval: 1m
            scrape_timeout: 10s
            evaluation_interval: 1m
          scrape_configs:
          - job_name: otel-collector
            honor_timestamps: true
            scrape_interval: 10s
            scrape_timeout: 10s
            metrics_path: /metrics
            scheme: http
            follow_redirects: true
            enable_http2: true
            http_sd_configs:
            - follow_redirects: false
              enable_http2: false
              url: http://allocator-targetallocator:80/jobs/otel-collector/targets?collector_id=$POD_NAME
        target_allocator:
          collector_id: ${POD_NAME}
          endpoint: http://allocator-targetallocator
          http_sd_config:
            refresh_interval: 60s
          interval: 30s
    service:
      extensions:
      - memory_ballast
      pipelines:
        metrics:
          exporters:
          - prometheusremotewrite
          processors:
          - memory_limiter
          - batch
          receivers:
          - prometheus
kind: ConfigMap
metadata:
  creationTimestamp: "2023-04-20T18:37:48Z"
  labels:
    app.kubernetes.io/component: opentelemetry-collector
    app.kubernetes.io/instance: otel.allocator
    app.kubernetes.io/managed-by: opentelemetry-operator
    app.kubernetes.io/name: allocator-collector
    app.kubernetes.io/part-of: opentelemetry
    app.kubernetes.io/version: latest
  name: allocator-collector
  namespace: otel
  ownerReferences:
  - apiVersion: opentelemetry.io/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: OpenTelemetryCollector
    name: allocator
    uid: 012622e4-db73-4c9c-8f7c-39e47cdabb8f
  resourceVersion: "9771170"
  uid: db4349e1-0f44-4b71-af3e-7467b80b7696
```

### PR DESCRIPTION
**Pull request by matej-g**
Resolves #1669.

This change adds logic to populate the so-called store assets - this includes information from service and pod monitors that relates to authorization information (e.g. basic auth, bearer token, TLS) - in order to propagate these information to scrape configs. Currently, this logic is not implement, resulting in scrape configurations missing proper credentials.

This PR also adds tests for the `LoadConfig` method for the Prometheus CR watcher.

### CODE REVIEW

File: cmd/otel-allocator/watcher/promOperator.go
Comment (jaronoff97) [inline cmd/otel-allocator/watcher/promOperator.go:197]: Just for posterity, could you include what _could_ be set here? (in case someone in the future asks why we can't do X with service monitors it makes it more discoverable)

File: cmd/otel-allocator/watcher/promOperator.go
Comment (jaronoff97) [inline cmd/otel-allocator/watcher/promOperator.go:227]: It looks like the method you linked is a public method, are we unable to use it?

Comment (jaronoff97) [reply cmd/otel-allocator/watcher/promOperator.go:227]: ah maybe because we don't want to make a whole resource selector. Interested in your perspective here :)

**Review by jaronoff97 — COMMENTED** [review]
Just a few questions... also, were you able to test this in a real cluster to confirm this works as expected?

File: cmd/otel-allocator/watcher/promOperator.go
Comment (matej-g) [reply cmd/otel-allocator/watcher/promOperator.go:227]: I was considering use those public methods, but eventually decided to borrow some of the code, because 1) as you said we would need to build the whole resource selector; 2) the methods for monitors seem to do extra stuff (whole selection and validation of monitors), whereas here we wanted to just populate the store.

However, now that you point this out, and as I noticed these methods were recently exported, it might be a good opportunity to replace this whole part of the watcher logic directly with the resource selector methods. We could get that extra validation etc. for free, but perhaps first I'd ensure those methods will result in correct selection of monitors for our purposes.

What do you think about doing this as a follow up work?

File: cmd/otel-allocator/watcher/promOperator.go
Comment (matej-g) [reply cmd/otel-allocator/watcher/promOperator.go:197]: Sure thing, I can add some TODO comments 👍

**Comment by matej-g** [comment]
@jaronoff97 forgot to add in the description - in addition to adding unit tests, I also tested this locally on a test cluster, with expected results, see resulting scrape config for an example service monitor with basic auth (checked in browser):

Screenshot 2023-05-04 at 16 18 33

**Comment by jaronoff97** [comment]
@matej-g awesome! how does this work with secret mounting? does the collector need to mount a secret volume or does the TA need to?

**Review by jaronoff97 — APPROVED** [review]
This LGTM. I want another review from @open-telemetry/operator-ta-maintainers before it's g2g.

File: cmd/otel-allocator/watcher/promOperator.go
Comment (matej-g) [reply cmd/otel-allocator/watcher/promOperator.go:224]: Let's do it 👍 I'll open an issue for this to ensure it won't go unaddressed.

**Review by Aneurysm9 — APPROVED** [review]

**Comment by jaronoff97** [comment]
@matej-g looks like a bad git merge 😅

**Comment by pavolloffay** [comment]
@matej-g could you please rebase?

**Comment by matej-g** [comment]
cc @open-telemetry/operator-ta-maintainers this should finally be good to merge, thank you 🙇

**Pull request merged**
